src/main/java/nwoolcan/controller/brewery/warehouse/WarehouseControllerImpl.java

```diff
@@ -0,0 +1,112 @@
+         this.warehouse = BreweryContext.getInstance().getWarehouse();
+    }
+
+    @Override
+    public List<AbstractArticleViewModel> getArticles(final QueryArticle queryArticle) {
+        return warehouse.getArticles(queryArticle)
+                        .stream()
+                        .map(AbstractArticleViewModel::getViewArticle)
+                        .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<AbstractStockViewModel> getStocks(final QueryStock queryStock) {
+        return warehouse.getStocks(queryStock)
+                        .stream()
+                        .map(AbstractStockViewModel::getViewStock)
+                        .collect(Collectors.toList());
+    }
+
+    @Override
+    public WarehouseViewModel getWarehouseViewModel() {
+        return new WarehouseViewModel(warehouse);
+    }
+
+    @Override
+    public MiscArticleViewModel createMiscArticle(final String name, final UnitOfMeasure unitOfMeasure) {
+        return new MiscArticleViewModel(warehouse.createMiscArticle(name, unitOfMeasure));
+    }
+
+    @Override
+    public BeerArticleViewModel createBeerArticle(final String name, final UnitOfMeasure unitOfMeasure) {
+        return new BeerArticleViewModel(warehouse.createBeerArticle(name, unitOfMeasure));
+    }
+
+    @Override
+    public IngredientArticleViewModel createIngredientArticle(final String name,
+                                                              final UnitOfMeasure unitOfMeasure,
+                                                              final IngredientType ingredientType) {
+        return new IngredientArticleViewModel(warehouse.createIngredientArticle(name,
+                                                                                unitOfMeasure,
+                                                                                ingredientType));
+    }
+
+    @Override
+    public Result<StockViewModel> createStock(final int articleId, final Date expirationDate) {
+        return Result.of(articleId)
+                     .flatMap(this::getArticleById)
+                     .flatMap(article -> warehouse.createStock(article, expirationDate))
+                     .map(StockViewModel::new);
+    }
+
+    @Override
+    public Result<StockViewModel> createStock(final int articleId) {
+        return Result.of(articleId)
+                     .flatMap(this::getArticleById)
+                     .flatMap(warehouse::createStock)
+                     .map(StockViewModel::new);
+    }
+
+    private Result<Article> getArticleById(final int articleId) {
+        final QueryArticle queryArticle = new QueryArticleBuilder().setMinID(articleId).setMaxID(articleId).build();
+        return Result.of(warehouse.getArticles(queryArticle))
+                     .require(articles -> articles.size() == 1, new IllegalArgumentException(ARTICLE_NOT_FOUND))
+                     .map(articles -> articles.get(0));
```

Comment:
not sure if can be improved

Reply:
you can use streams and call `findFirst()` if you like it more